### PR TITLE
Fix Windows tests and add Windows CI

### LINKED ISSUE #250: Fix Windows tests
Currently, many tests fail on Windows. #247 solves most of them, but two tests, `testBadCharacters` and `testBuildHeaderGuardWithRepository` still don't work on Windows and have hence been skipped.

### PR DESCRIPTION
**Pull request by jayvdb**
https://github.com/cpplint/cpplint/pull/248 is first in this series.

### CODE REVIEW

**Comment by jayvdb** [comment]
Tests are hanging on some windows versions of Python.

File: tox.ini
Comment (jayvdb) [inline tox.ini]: WIP - `testBadCharacters` is the test which is causing Windows to hang.

**Comment by aaronliu0130** [comment]
What would happen if the envlist in tox only contains the three envs we want?

**Comment by aaronliu0130** [comment]
You don't need to force push; you can just click on the dropdown next to "merge" and click on "squash and merge" when everything's finished.

**Comment by aaronliu0130** [comment]
Maybe we should only test 3.7 and 3.12 on Windows? pypy is slow and compatibility-wise, I don't see any issues with 3.10 that wouldn't be caught by the ubuntu runner or the other versions ran on Windows.

File: .github/workflows/ci.yml
Comment (aaronliu0130) [inline .github/workflows/ci.yml:1]: Maybe we should only test 3.7 and 3.12 on Windows? pypy is slow and compatibility-wise, I don't see any issues with 3.10 that wouldn't be caught by the ubuntu runner or the other versions ran on Windows.

Comment (jayvdb) [reply .github/workflows/ci.yml:1]: CI Testing of the support grid is *more* important for Windows, even if it is slow, because it is less likely that PRs submitted will have been tested on Windows.

Comment (aaronliu0130) [reply .github/workflows/ci.yml:1]: Then, should we disable it for Ubuntu?

Comment (jayvdb) [reply .github/workflows/ci.yml:1]: What would be the benefit of disabling CI of pypy on Linux?

Comment (aaronliu0130) [reply .github/workflows/ci.yml:1]: Shortening the time it takes for workflows to finish. IMO there is marginal benefit to having two pypy tests.

Comment (jayvdb) [reply .github/workflows/ci.yml:1]: The Ubuntu one always finishes before the Windows one.
Removing the Ubuntu one doesnt speed up the Windows one (it can speed up other Ubuntu ones, but that only occurs on much larger matrix, and only matters if the project has lots of contributors submitting lots of PRs per day)

Comment (aaronliu0130) [reply .github/workflows/ci.yml:1]: Huh, I was under the impression that both had similar times. But since the Windows one takes almost twice the time, I think we should remove it defer to issue reports if any such breakages occur. Such platform-specific breakages should still be rare enough.

Comment (jayvdb) [reply .github/workflows/ci.yml:1]: As you can see in this PR, there is a Windows specific bug that needs to be fixed.  We need CI to verify it is fixed on all supported Windows versions.

Testing min, max & pypy on Windows is common.  That is what we are doing.  6mins for CI to finish is extremely fast.  See for example https://github.com/PyCQA/pyflakes , https://github.com/amperser/proselint , https://github.com/sqlfluff/sqlfluff , https://github.com/jorisroovers/gitlint .  If it was over 15mins, then some discussions around how to optimise CI is worthwhile, but it is not uncommon for CI to take an hour .  Usually the more mature a project, the longer CI takes, because the maintainers have learnt to use CI to prevent the uncommon breakages.

Comment (aaronliu0130) [reply .github/workflows/ci.yml:1]: Nearly all Windows-specific bugs can be seen in the tests for 3.7 and 3.12. If someone indeed gets a fail on pypy on Windows, they can file an issue. It doesn't seem like something worth enough, especially for such a relatively small project like this where fixes can be faster.

Comment (tkruse) [reply .github/workflows/ci.yml:1]: Personally I believe avoiding breakage in releases is much more valuable than any time saved in release CI. In local development, it may make sense to use a different tox setup with less envs tested, but release CI should be as diligent as technically feasible. Any interruption, even if rare, costs both the users and maintainers much more effort and frustration than waiting times on CI.

If there is a way, possibly merges to develop could have quicker CI than merges to master, as it's the releases we need to protect, but not every little change to develop. Basically any solution that still uses all envs **for release** is fine with me.

(This is different between projects, for some projects I would argue that CI time needs to be optimized to be shorter, so I am not making a strong statement about all software development. This is about cpplint. With cpplint, maintainers might be unavailable to respond to requests for months at a time, there is no 24h support)

Comment (aaronliu0130) [reply .github/workflows/ci.yml:1]: Pretty sure we'd need a lot of merge commits to master if we had different CIs for master and develop. That was one of the reasons master currently has weird commit history.

Comment (tkruse) [reply .github/workflows/ci.yml:1]: I don't know the solution, but I would expect that github can run different actions for PRs targeting master or other branches. If not, then I'd rather keep slow CI that we can trust over fast CI with gaps.

Comment (aaronliu0130) [reply .github/workflows/ci.yml:1]: AFAIK, the best way to do that would be to add a separate job (inside the same workflow) that generates the matrix.

However, since all commits to develop require a pull request now, I feel like this is moot.

Comment (jayvdb) [reply .github/workflows/ci.yml:1]: I agree this is moot.  There will only be one active branch for development.

Can I get an approval of this PR please.

**Review by tkruse — APPROVED** [review]

**Comment by aaronliu0130** [comment]
@jayvdb Could we make pytest output unverbose again? It's currently a bit more of a pain to navigate in CI's output without much benefit that I can see.

**Comment by jayvdb** [comment]
Because when tests *hang* in CI, pytest doesnt show which test has hung.  The best way to see which test hung is to know which test was previous to the hang.

**Pull request merged**
